File: micronaut-mcp-server-java-sdk/src/test/java/io/micronaut/mcp/server/utils/PgnLoader.java
package io.micronaut.mcp.server.utils;

import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.io.ResourceLoader;
import jakarta.inject.Singleton;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Optional;

@Singleton
public class PgnLoader {
    private static final Logger LOG = LoggerFactory.getLogger(PgnLoader.class);
    private final ResourceLoader resourceLoader;

    PgnLoader(ResourceLoader resourceLoader) {
        this.resourceLoader = resourceLoader;
    }

    @NonNull
    public Optional<String> loadPgn(@NonNull @NotNull @Positive Integer round) {
        Optional<InputStream> roundPgnInputStreamOptional = resourceLoader.getResourceAsStream("classpath:fidewwc2024/round_" + round + ".pgn");
        if (roundPgnInputStreamOptional.isEmpty()) {
            return Optional.empty();
        }
        InputStream inputStream = roundPgnInputStreamOptional.get();
        try {
            return Optional.of(new String(inputStream.readAllBytes(), StandardCharsets.UTF_8));
        } catch (IOException e) {
            if (LOG.isErrorEnabled()) {
                LOG.error(e.getMessage(), e);
            }
            return Optional.empty();
        }
    }
}

File: test-suite-docs-python/src/test/python/micronaut/mcp/docs/resources/PgnLoader.py
import logging

from jakarta.inject import Singleton
from java.lang import String
from java.nio.charset import StandardCharsets
from micronaut.core.io import ResourceLoader

LOG = logging.getLogger(__name__)


@Singleton
class PgnLoader:
    """Loads the PGN of a round of the World Chess Championship 2024 from the classpath."""

    def __init__(self, resource_loader: ResourceLoader):
        self.resource_loader = resource_loader

    def load_pgn(self, round: int) -> str | None:
        pgn_input_stream = self.resource_loader.getResourceAsStream(f"classpath:fidewwc2024/round_{round}.pgn")
        if pgn_input_stream.isEmpty():
            return None
        input_stream = pgn_input_stream.get()
        try:
            return str(String(input_stream.readAllBytes(), StandardCharsets.UTF_8))
        except Exception as e:
            LOG.error("%s", e, exc_info=e)
            return None
        finally:
            input_stream.close()
